services/repository/files/content.go

```diff
@@ -217,15 +218,15 @@ func GetContents(ctx context.Context, repo *repo_model.Repository, treePath, ref
 	}
 	// Handle links
 	if entry.IsRegular() || entry.IsLink() {
-		downloadURL, err := url.Parse(fmt.Sprintf("%s/raw/%s/%s/%s", repo.HTMLURL(), refType, ref, treePath))
+		downloadURL, err := url.Parse(fmt.Sprintf("%s/raw/%s/%s/%s", repo.HTMLURL(), refType, ref, util.PathEscapeSegments(treePath)))
```

Comment:
```suggestion
		downloadURL, err := url.Parse(fmt.Sprintf("%s/raw/%s/%s/%s", repo.HTMLURL(), url.PathEscape(refType), util.PathEscapeSegments(ref), util.PathEscapeSegments(treePath)))
```

Reply:
Does it work if `ref` contains `/` ? eg: ref="a/b", treePath="c/d", then there will be "/a/b/c/d", which is the ref part.

Reply:
The ref should be escape path segments.

The code does the magic of handling the ref

Reply:
Oh yup, that's `getRefNameFromPath`. By  design if ref `/a`  exists, there should be no ref `/a/b`, so it's safe.

Reply:
I think the ref name should always be URL safe but... It's really much better to just escape the thing